### PR TITLE
add getMaterialPropertyDot

### PR DESCRIPTION
**Pull request by YaqiWang**
For solving  #12153. This is just an idea.

### CODE REVIEW

File: framework/src/materials/MaterialData.C [hunk 1]
```diff
@@ -70,6 +71,50 @@ MaterialData::reinit(const std::vector<std::shared_ptr<Material>> & mats)
...
+MaterialData::reinitDotProps(Real dt, Real dt_old, Moose::TimeIntegratorType scheme)
+{
+  for (unsigned int i = 0; i < _dot_props.size(); ++i)
+  {
+    MaterialProperty<Real> & dot_prop = *_dot_props[i];
+    const MaterialProperty<Real> & old_prop = *_old_props[i];
+    const MaterialProperty<Real> & older_prop = *_older_props[i];
+    const MaterialProperty<Real> & prop = *_current_props[i];
+
+    switch (scheme)
```
Comment (friedmud) [inline framework/src/materials/MaterialData.C]: This isn't the right place for this.  If you add a new time integrator... the properties won't get integrated properly.

**Review by friedmud — CHANGES_REQUESTED** [review]
The time derivative calculations are hardcoded in the wrong place - they should be part of the TimeIntegrator class.

File: framework/src/materials/MaterialData.C (hunk 1 above)
Comment (YaqiWang) [reply framework/src/materials/MaterialData.C]: The problem is that this evaluation has to go with material property evaluations. Time integrators do not provide a place for holding this calculation though.

Comment (YaqiWang) [reply framework/src/materials/MaterialData.C]: We can add a new function in `TimeIntegrator`, which does this evaluation and can be called in `FEProblemBase::reinitMaterials`. Let me change this PR a little.

**Comment by YaqiWang** [comment]
Well, my documentation in `framework/doc/content/syntax/Materials/index.md` appears not following the standard. Also it is horribly incomplete (not my fault though). I do not quite like the fact that `getMaterialPropertyDot` only supports `Real` type, but do not know how we can go further. @permcody I think this is ready for review. A moose developer needs to document our general ideas on how materials are implemented.

**Comment by permcody** [comment]
OK - thanks! I realize that a lot of the base systems don't have documentation yet. We are working on it, but it takes time. Thanks for at least making an attempt to get it started.

File: framework/src/materials/MaterialPropertyInterface.C [hunk 2]
```diff
@@ -196,3 +196,40 @@ MaterialPropertyInterface::checkExecutionStage()
...
+               name,
+               "\" was requested.");
+
+  // Check if the supplied parameter is a valid input parameter key
+  std::string prop_name = deducePropertyName(name);
+
+  // Check if it's just a constant
+  const MaterialProperty<Real> * default_property = defaultMaterialProperty<Real>(prop_name);
+  if (default_property)
+    return *default_property;
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyInterface.C]: If it's constant then you should return `0`.

File: framework/include/materials/MaterialData.h [hunk 3]
```diff
@@ -161,6 +176,12 @@ class MaterialData
   /// Status of storage swapping (calling swap sets this to true; swapBack sets it to false)
   bool _swapped;
 
+  /// Properties having dot properties
+  std::vector<const MaterialProperty<Real> *> _current_props;
+  std::vector<const MaterialProperty<Real> *> _old_props;
+  std::vector<const MaterialProperty<Real> *> _older_props;
+  std::vector<MaterialProperty<Real> *> _dot_props;
```
Comment (friedmud) [inline framework/include/materials/MaterialData.h]: How are these different from `_props_old` above?

File: framework/include/materials/MaterialData.h [hunk 4]
```diff
@@ -161,6 +176,12 @@ class MaterialData
   /// Status of storage swapping (calling swap sets this to true; swapBack sets it to false)
   bool _swapped;
 
+  /// Properties having dot properties
+  std::vector<const MaterialProperty<Real> *> _current_props;
+  std::vector<const MaterialProperty<Real> *> _old_props;
+  std::vector<const MaterialProperty<Real> *> _older_props;
```
Comment (lindsayad) [inline framework/include/materials/MaterialData.h]: So we have to basically add three data members that conceptually mimic the members 10 lines earlier?

File: framework/doc/content/syntax/Materials/index.md [hunk 5]
```diff
@@ -1,4 +1,12 @@
-<!-- MOOSE Documentation Stub: Remove this when content is added. -->
+# Materials System Overview
+
+Time derivative of a material property could be needed for transient calculations.
+MOOSE provides a function getMaterialPropertyDot that can be called in kernels, bcs, user objects, etc,
+which returns a reference of a time derivative of a real scalar material property.
+The time derivative obtained with this function is consistent with the time integration scheme used by the calculation.
+It is noted that calling this function will make the material property stateful.
+MOOSE will keep three copies of the property on all quadrature points for the current, old and older
+properties that are used for evaluating the time derivative.
```
Comment (friedmud) [inline framework/doc/content/syntax/Materials/index.md]: Please add an even larger "note" about it being for scalar (`Real`) material properties only.  Use one of these: http://mooseframework.org/utilities/MooseDocs/extensions/alert.html

File: framework/src/timeintegrators/TimeIntegrator.C [hunk 6]
```diff
@@ -69,3 +70,22 @@ TimeIntegrator::getNumLinearIterationsLastSolve() const
...
 }
+
+void
+TimeIntegrator::computeDotProperties(Moose::MaterialDataType type, THREAD_ID tid) const
+{
+  auto material_data = _fe_problem.getMaterialData(type, tid);
+
+  auto props = material_data->getPropertiesForDot();
+  auto old_props = material_data->getOldPropertiesForDot();
+  auto dot_props = material_data->getDotProperties();
```
Comment (friedmud) [inline framework/src/timeintegrators/TimeIntegrator.C]: 😱 those `auto`s should be `auto &` (or `const auto &` if you can).  You are making copies of these vectors here!  Be careful!

File: framework/src/timeintegrators/BDF2.C [hunk 7]
```diff
@@ -64,6 +65,28 @@ BDF2::computeTimeDerivatives()
...
 
+void
+BDF2::computeDotProperties(Moose::MaterialDataType type, THREAD_ID tid) const
+{
+  auto material_data = _fe_problem.getMaterialData(type, tid);
+
+  auto props = material_data->getPropertiesForDot();
+  auto old_props = material_data->getOldPropertiesForDot();
+  auto older_props = material_data->getOlderPropertiesForDot();
+  auto dot_props = material_data->getDotProperties();
```
Comment (friedmud) [inline framework/src/timeintegrators/BDF2.C]: All of these should be `auto &`... you are making vector copies!

File: framework/src/timeintegrators/CrankNicolson.C [hunk 8]
```diff
@@ -38,6 +38,25 @@ CrankNicolson::computeTimeDerivatives()
...
 }
 
+void
+CrankNicolson::computeDotProperties(Moose::MaterialDataType type, THREAD_ID tid) const
+{
+  auto material_data = _fe_problem.getMaterialData(type, tid);
+
+  auto props = material_data->getPropertiesForDot();
+  auto old_props = material_data->getOldPropertiesForDot();
+  auto dot_props = material_data->getDotProperties();
```
Comment (friedmud) [inline framework/src/timeintegrators/CrankNicolson.C]: All of these should be `auto &` you're making vector copies!

File: framework/include/materials/MaterialData.h (hunk 4 above)
Comment (dschwen) [reply framework/include/materials/MaterialData.h]: And `_props_old` vs. `_old_props` ?! The new names offer no indication of what sets these members apart from the properties above.

File: framework/src/materials/MaterialPropertyInterface.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/materials/MaterialPropertyInterface.C]: oh, right!

File: framework/include/materials/MaterialData.h (hunk 4 above)
Comment (YaqiWang) [reply framework/include/materials/MaterialData.h]: These three members only contains properties in REAL that are to be used for evaluating dot properties. They are used in time integrators. We can have a vector of bools in size of `_props.size()` to indicate if the properties are going to be used for dot evaluation. We will have to cast the property into real later.

File: framework/src/timeintegrators/TimeIntegrator.C [hunk 9]
```diff
@@ -69,3 +70,22 @@ TimeIntegrator::getNumLinearIterationsLastSolve() const
...
+
+  for (unsigned int i = 0; i < dot_props.size(); ++i)
```
Comment (friedmud) [inline framework/src/timeintegrators/TimeIntegrator.C]: Use range-based `for`

File: framework/src/timeintegrators/BDF2.C [hunk 10]
```diff
@@ -64,6 +65,28 @@ BDF2::computeTimeDerivatives()
...
+
+  for (unsigned int i = 0; i < dot_props.size(); ++i)
```
Comment (friedmud) [inline framework/src/timeintegrators/BDF2.C]: range-based `for`

File: framework/src/timeintegrators/CrankNicolson.C [hunk 11]
```diff
@@ -38,6 +38,25 @@ CrankNicolson::computeTimeDerivatives()
...
+
+  for (unsigned int i = 0; i < dot_props.size(); ++i)
```
Comment (friedmud) [inline framework/src/timeintegrators/CrankNicolson.C]: range-based `for`

**Review by friedmud — CHANGES_REQUESTED** [review]
I do wish we could do this for any type (not only `Real`) but this is a good start!

File: framework/src/timeintegrators/BDF2.C (hunk 7 above)
Comment (YaqiWang) [reply framework/src/timeintegrators/BDF2.C]: My misunerstanding, thought `auto` will respect `&`.

File: framework/src/timeintegrators/BDF2.C (hunk 10 above)
Comment (YaqiWang) [reply framework/src/timeintegrators/BDF2.C]: need `i` to index the other three vectors.

File: framework/include/materials/MaterialData.h (hunk 3 above)
Comment (YaqiWang) [reply framework/include/materials/MaterialData.h]: They contains pointers of material properties that are only needed for dot evaluation.

File: framework/include/materials/MaterialData.h (hunk 4 above)
Comment (friedmud) [reply framework/include/materials/MaterialData.h]: That sounds better

**Comment by YaqiWang** [comment]
OK, a new implementation. I think this is better.

File: framework/src/materials/MaterialPropertyInterface.C [hunk 12]
```diff
@@ -88,6 +88,31 @@ MaterialPropertyInterface::defaultMaterialProperty(const std::string & name)
   return nullptr;
 }
 
+const MaterialProperty<Real> *
+MaterialPropertyInterface::zeroMaterialPropertyDot(const std::string & name)
```
Comment (dschwen) [inline framework/src/materials/MaterialPropertyInterface.C]: This should not have a `Dot` in its name. It is more general than just for dot properties. I'm already doing a similar thing for the dervative material properties and might just used this function instead.

**Comment by YaqiWang** [comment]
Comments are addressed. Calling for new review. Thanks.

File: framework/src/materials/MaterialPropertyInterface.C [hunk 13]
```diff
@@ -88,6 +88,31 @@ MaterialPropertyInterface::defaultMaterialProperty(const std::string & name)
...
+{
+  std::istringstream ss(name);
+  Real real_value;
+
+  // check if the string parsed cleanly into a Real number
+  if (ss >> real_value && ss.eof())
```
Comment (dschwen) [inline framework/src/materials/MaterialPropertyInterface.C]: My suggestion would be to factor this check out into the caller and name this method ` zeroMaterialProperty` without the `Dot`. There are applications like the derivative material system that need to generate and return zero properties. I have code that does that already, but may want to refactor my code to use this new function.

Comment (YaqiWang) [reply framework/src/materials/MaterialPropertyInterface.C]: I actually just noticed that there is a function `getZeroMaterialProperty`. I should just use that. Checking constant value will be in `getMaterialPropertyDot`.

File: framework/doc/content/syntax/Materials/index.md [hunk 14]
```diff
@@ -1,4 +1,19 @@
...
+
+Time derivative of a material property could be needed for transient calculations.
+MOOSE provides a function getMaterialPropertyDot that can be called in kernels, bcs, user objects, etc,
+which returns a reference of a time derivative of a real scalar material property.
+
+!alert note
+getMaterialPropertyDot currently only supports real scalar (Real) material property.
+
+!alert note
+getMaterialPropertyDot is not available in Materials.
```
Comment (YaqiWang) [inline framework/doc/content/syntax/Materials/index.md]: I will need to override this function in `Material.h` to throw an error when users are calling this in `Material`. I did not realize `Material` is also derived from `MaterialPropertyInterface`.

File: framework/src/materials/MaterialPropertyInterface.C (hunk 13 above)
Comment (YaqiWang) [reply framework/src/materials/MaterialPropertyInterface.C]: So @dschwen I guess `getZeroMaterialProperty` will work for you and you will not need `zeroMaterialProperty`, right?

Comment (dschwen) [reply framework/src/materials/MaterialPropertyInterface.C]: Given that I wrote `getZeroMaterialProperty`... yes.

**Comment by friedmud** [comment]
@YaqiWang I was looking at your tests - have you done tests to actually verify that you are getting the correct rate of convergence for your problem with each of the time integrators?

**Comment by YaqiWang** [comment]
No, I have not. We can set up a test solving `d(u/t)/dt = -c u`, where we use material property for the `t` under `u`. It is equivalent with `du/dt = -(c*t - 1/t)u`, which can be solved without `getMaterialPropertyDot`. I am in middle of something else thus hesitate spending time on doing this investigation.

**Comment by friedmud** [comment]
Ok - no hurry - but I kind of don't want to merge this until that testing is complete.  If we're going to provide a time-derivative that we claim is accurate... then we need to really know that it is.

Not being a hard-ass here... we just need to be careful with this stuff!

**Comment by YaqiWang** [comment]
Ok, I will get to this once I finish my stuff. Meanwhile, if you have any further comments on this PR, let me know.

**Comment by YaqiWang** [comment]
Forgot mentioning, function specialization of a template class seems can make `getMaterialPropertyDot` fully templated. We just need to add some specializations for supported types, like `Real`, `RealVectorValue`, RealTensorValue`, `std::vector<Real>`, etc. I can try that out when I get back on this.

**Comment by YaqiWang** [comment]
I must ignore something important, in my last commit, running the input `dot_convergence_study_nodot.i`, which does not use `getMaterialPropertyDot` at all, I am not getting the correct convergence order for Crank-Nicolson scheme. I copied @jwpeterson 's script and made some changes to run the input. I will take a look tomorrow, but I'd like to get your feedback what could be wrong in `dot_convergence_study_nodot.i`.

**Comment by YaqiWang** [comment]
Looks like any time dependent quantity cannot be included in time kernel, otherwise the expected convergence rate is destroyed. The convergence with `dot_convergence_study_nodot.i` is
[destroyed.pdf](https://github.com/idaholab/moose/files/2454672/destroyed.pdf)


Applying the following change:
```
diff --git a/test/tests/materials/dot_prop/dot_convergence_study_nodot.i b/test/tests/materials/dot_prop/dot_convergence_study_nodot.i
index 74af8df077..d95f6c179c 100644
--- a/test/tests/materials/dot_prop/dot_convergence_study_nodot.i
+++ b/test/tests/materials/dot_prop/dot_convergence_study_nodot.i
@@ -19,8 +19,6 @@
   [./time]
     type = CoefTimeDerivative
     variable = u
-    mat_prop_coef = time_prop
-    prop_in_derivative = false
   [../]
   [./reaction]
     type = CoefReaction
@@ -44,7 +42,7 @@
 [Functions]
   [./time_function]
     type = ParsedFunction
-    value = 't'
+    value = '1/t'
   [../]
 []
```
the convergence plot becomes
[plot.pdf](https://github.com/idaholab/moose/files/2454670/plot.pdf)
This may make this `getMaterialPropertyDot` completely unnecessary and it is better to move the time-dependent quantities into right hand side...

**Comment by dschwen** [comment]
@YaqiWang I largely agree. Similarly to the way fe calculate gradients of material properties using the chain rule and utilizing the gradients of the variables teh material property depends on, we need to use the dots of the non-linear variables the heat capacity depends on!

We do not support gradients of material properties, and in the same vein we should not support time derivatives of material properties. 

As a consequence, in these applications the material properties should not _directly_ depend on _t_, but should only depend on variables that may have a time dependence.

**Comment by YaqiWang** [comment]
Material properties depending on *t* directly or indirectly should be fine unless it is not in the time kernel. If the property is outside `du/dt`, it creates the time convergence issue; if it is inside like `d(p*u)/dt`, chain rule will add an approximation and we better define `p*u` to a new variable. I like this *We do not support gradients of material properties, and in the same vein we should not support time derivatives of material properties.*.

**Comment by friedmud** [comment]
@YaqiWang - right.  Just got ask @rcm59 about it - he hates the idea that we ever use rho * Cp * dT/dt for heat conduction!  It's only true if rho and Cp are constant in time (which they almost never are because they typically depend on T).

Rich would always rather do stuff in full conservative mode: d (rho * Cp * T) / dt .... and of course that's what we do in fluid mechanics (where this really matters).

BUT: For _most_ heat conduction problems rho and Cp vary _slowly_ with time... and therefore it's not a big deal (but will show up in a time convergence study!).  Might not be true for some reactor problems though!

**Comment by dschwen** [comment]
@YaqiWang , what I was getting at above is that to me it seems that

`d(rho*Cp*T)/dt = drho/dt * Cp * T + rho * dCp/dt * T + rho * Cp * dT/dt`

where

`drho(T)/dt = drho/dT*dT/dt`
`dCp(T)/dt = dCp/dT*dT/dt`

(and accordingly for dependencies on further time dependent variables). We can already calculate those, can't we?

**Comment by YaqiWang** [comment]
No @dschwen, internal energy E(T) is defined as `\int_0^T rho Cp(T') dT'` thus `dE/dt = dE/dT* dT/dt`, which is `rho Cp dT/dt`. I do not know how thermal expansion get into the term but assuming no expansion. `rho Cp` in the time kernel will cause the convergence issue, for example with Crank-Nicolson. If it is in time kernel, we possibly do not have a clean way of consider it as it is in right hand side.

**Pull request closed without merging**